### PR TITLE
Bump React to v16 alpha

### PR DESCRIPTION
**Pull request by aduth**
This pull request seeks to update our React dependency to track the latest alpha version, currently `16.0.0.9-alpha.9`. Doing so will allow us to stay ahead of the curve in anticipation of upcoming changes. As the plugin becomes more widely distributed, we'll want to shift this back to a less unstable version, but I'd argue it's fine to opt in to an alpha dependency for our alpha plugin (we're also using the TinyMCE nightly version).

There appear to be no breaking changes with this update, though it comes with a few additions we'll want to take advantage of. The ["Fiber" reconciliation algorithm](https://gist.github.com/duivvv/2ba00d413b8ff7bc1fa5a2e51c61ba43) is the main headline for this release, but for our purposes, we're mostly interested in newly supported return values for components: strings and arrays. This will help, in https://github.com/WordPress/gutenberg/pull/409#issuecomment-294732502 for example, in avoiding unnecessary wrapper elements in a `save` return value.

### CODE REVIEW

File: index.php [hunk 1]
```diff
@@ -35,8 +35,9 @@ function gutenberg_register_scripts() {
 	$suffix = SCRIPT_DEBUG ? '' : '.min';
 
 	// Vendor
-	wp_register_script( 'react', 'https://unpkg.com/react@15/dist/react' . $suffix . '.js' );
-	wp_register_script( 'react-dom', 'https://unpkg.com/react-dom@15/dist/react-dom' . $suffix . '.js', array( 'react' ) );
+	$react_suffix = ( SCRIPT_DEBUG ? '.development' : '' ) . $suffix;
```
Comment (youknowriad) [inline index.php]: Should we fallback to `.production` to avoid errors

**Review by youknowriad — COMMENTED** [review]
Feel free to merge this when the `production` script is fixed

File: index.php (hunk 1 above)
Comment (aduth) [inline index.php]: >Should we fallback to `.production` to avoid errors

Yes, added in e25e732.

**Comment by aduth** [comment]
There's now an unmet peer dependency from `react-textarea-autosize` which has stagnated for a while now. Was the reason we have React in `package.json` because of this dependency? We could leave it as the old version if so. In general its presence there is a little confusing.

**Comment by youknowriad** [comment]
Maybe we could just drop this dependency for now (or replace it with our own implementation)?

**Comment by aduth** [comment]
I bumped the `package.json` back down to latest, since it's clearly only necessary for dependencies which define it as a required peer. The Webpack `externals` configuration should still ensure that it references the window global (16+). It wasn't entirely necessary, but I swapped `react-autosize-textarea` with `react-textarea-autosize` since it seems to be more maintained (at least resolved PropTypes warning).

**Pull request merged**